Fix handling indices stats request when all shards are missing

areek opened the pull request:
Currently, when an index exists in the cluster state but has no shards for reporting stats,
the missing stats object cause a `NullPointerException` when requesting the indices stats.
In this commit missing stats object for an index are initialized as empty stats instead
of null, honoring the stats flags set in the stats request. The commit fixes the issue for all
APIs that use the indices stats API namely `_cat/indices`, `_cat/shards` and `_stats`.

closes #20298

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/stats/ShardStats.java:
```diff
@@ -54,6 +54,30 @@ public ShardStats(ShardRouting routing, ShardPath shardPath, CommonStats commonS
+                primaryFound = true;
+            }
+        }
+        return primaryFound ? primaryStats : new CommonStats(CommonStatsFlags.ALL);
```
why the all flags? it seems arbitrary?

bleskes commented:
If I understand correctly the problem is that it may be the case that an index exists but it doesn't have any shard causing stats elements to be missing. I agree it would be nice to return empty stats objects rather than null, but I think we should honor the initial flags rather than blindly supply all stats.

also can we have tests?

javanna commented:
I don't see anything specific to cat indices among these changes, I think they modify indices stats api, hence also cat indices that uses it. Can we update the title/description/labels of the PR to reflect what apis it changes?

javanna commented:
One more thing: I wonder if this issue is specific to the tribe node, or can also happen with any node, or maybe just client nodes?

areek commented:
> If I understand correctly the problem is that it may be the case that an index exists but it doesn't have any shard causing stats elements to be missing. I agree it would be nice to return empty stats objects rather than null, but I think we should honor the initial flags rather than blindly supply all stats.

Thanks @bleskes for clarifying the problem statement, I was investigating this issue with tribe node, hence was too specific in my description. Now we honor the request stats flags instead of supplying all the stats by carrying over the request stats flag to the stats response.  

> also can we have tests?

Added this condition in `RestIndicesActionTests`, I wanted to add unit tests for this but it turned out to be non-trivial. 

> Can we update the title/description/labels of the PR to reflect what apis it changes?

Thanks @javanna for pointing it out, I updated these accordingly.

>  I wonder if this issue is specific to the tribe node, or can also happen with any node, or maybe just client nodes?

After some digging, this issue can happen on any node, when indices are removed/closed after the indices have been resolved for the stats request and before executing the stats request on the node.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

dakrone commented:
@areek is thing still being worked on?

rjernst commented:
I'm closing this PR since I do not think @areek is interested in bringing it up to date with master. If that is incorrect, please reopen.

clintongormley commented:
This is still a bug that needs to be fixed - I'll mark it as adoptme

zcola commented:
2.4.4 will still appear, we use "_cat / indices" as node health monitoring，Not only tribe node, but also data node will appear

rjernst commented:
@clintongormley I'm going to close this again, as we already have an issue to track this (#20298) and anyone who wants address this will still need to open a new PR.
